src/queue/multilevel.rs

```diff
@@ -0,0 +1,859 @@
+        let running_time = extras.running_time.clone();
+        let level = extras.current_level;
+        let begin = Instant::now();
+        let res = self.inner.handle(spawn, task_cell);
+        let elapsed = begin.elapsed();
+        if let Some(running_time) = running_time {
+            running_time.inc_by(elapsed);
+        }
+        if level == 0 {
+            self.manager.level0_elapsed.inc_by(elapsed);
+        }
+        let current_total = Duration::from_micros(
+            self.manager.total_elapsed.inc_by(elapsed) + elapsed.as_micros() as u64,
+        );
+        if current_total > ADJUST_CHANCE_INTERVAL {
+            self.manager.maybe_adjust_chance();
+        }
+        res
+    }
+
+    fn pause(&mut self, spawn: &mut Self::Spawn) -> bool {
+        self.inner.pause(spawn)
+    }
+
+    fn resume(&mut self, spawn: &mut Self::Spawn) {
+        self.inner.resume(spawn)
+    }
+
+    fn end(&mut self, spawn: &mut Self::Spawn) {
+        self.inner.end(spawn)
+    }
+}
+
+struct LevelManager {
+    level0_elapsed: ElapsedTime,
+    total_elapsed: ElapsedTime,
+    task_elapsed_map: TaskElapsedMap,
+    level_time_threshold: [Duration; LEVEL_NUM - 1],
+    level0_chance: AtomicU32,
+    level0_proportion_target: f64,
+    adjusting: AtomicBool,
+}
+
+impl LevelManager {
+    fn prepare_before_push<T>(&self, task_cell: &mut T)
+    where
+        T: TaskCell,
+    {
+        let extras = task_cell.mut_extras();
+        let task_id = extras.task_id;
+        let running_time = extras
+            .running_time
+            .get_or_insert_with(|| self.task_elapsed_map.get_elapsed(task_id));
+        let current_level = match extras.fixed_level {
+            Some(level) => level,
+            None => {
+                let running_time = running_time.as_duration();
+                self.level_time_threshold
+                    .iter()
+                    .enumerate()
+                    .find(|(_, &threshold)| running_time < threshold)
+                    .map(|(level, _)| level)
+                    .unwrap_or(LEVEL_NUM - 1) as u8
+            }
+        };
+        extras.current_level = current_level;
+        extras.schedule_time = Some(now());
+    }
+
+    fn get_level0_chance(&self) -> u32 {
+        self.level0_chance.load(SeqCst)
+    }
+
+    fn maybe_adjust_chance(&self) {
+        if self.adjusting.compare_and_swap(false, true, SeqCst) {
+            return;
+        }
+        // The statistics may be not so accurate because we cannot load two
+        // atomics at the same time.
+        let total = self.total_elapsed.0.load(SeqCst);
+        if Duration::from_micros(total) < ADJUST_CHANCE_INTERVAL {
+            // Another thread just adjusted the chances.
+            return;
+        }
+        let total = self.total_elapsed.0.swap(0, SeqCst);
+        let level0 = self.level0_elapsed.0.swap(0, SeqCst);
+        let current_proportion = level0 as f64 / total as f64;
```

Comment:
What if total is zero?

Reply:
`total` must be at least `ADJUST_CHANCE_INTERVAL` thus cannot be zero.

Reply:
But it just gets swapped at L300.

Reply:
`maybe_adjust_chance` is locked by `adjusting`, so there is only one thread executing this function. `total_elapsed` increases monotonously except the swap at L300. It means the total at L300 is not less than the total at L295 so it cannot be zero.